metadata/src/utils/retain.rs

```diff
@@ -261,13 +404,62 @@ pub fn retain_metadata<F, G>(
         .iter()
         .find(|ty| ty.ty.path.segments == ["sp_runtime", "DispatchError"])
         .expect("Metadata must contain sp_runtime::DispatchError");
-    type_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(dispatch_error_ty.id);
+    type_set.seen_ids.insert(metadata.runtime_ty);
+    type_set.collect_extrinsic_types(&metadata.extrinsic);
+
+    // Collect the outer enums type IDs.
+    for typ in [
+        metadata.outer_enums.call_enum_ty,
+        metadata.outer_enums.event_enum_ty,
+        metadata.outer_enums.error_enum_ty,
+    ] {
+        type_set.collect_variants_in_type(metadata, typ, &mut pallets_filter);
+    }
 
-    // Now, keep the type IDs we've asked for. This recursively keeps any types referenced from these.
-    // This will return a map from old to new type ID, because IDs may change.
-    let map_ids = metadata.types.retain(|id| type_ids.contains(&id));
+    let mut retained_set = TypeSet {
+        seen_ids: type_set
+            .seen_ids
+            .difference(&retained_set.seen_ids)
+            .copied()
+            .collect(),
+    };
+    // Strip down Pallets we dont need and only keep types that are not yet included in the retained set.
+    for pallet in metadata.pallets.values_mut() {
+        if !pallets_filter(&pallet.name) {
+            type_set.update_filtered_pallet(pallet, &mut retained_set);
+        }
+    }
+
+    // Filter out unnecesary pallets that have no entries
+    metadata.pallets.retain(|pallet| {
+        pallets_filter(&pallet.name)
+            || !matches!(
+                pallet,
+                PalletMetadataInner {
+                    storage: None,
+                    call_ty: None,
+                    event_ty: None,
+                    error_ty: None,
+                    constants: map,
+                    ..
+                } if map.is_empty()
```

Comment:
Does this make any difference? I can't see why a pallet would exist that had nothing meaningful in it, but even if it did, is there any harm in leaving it there?

Reply:
Ah I am starting to see why this is kept but not understanding it yet; you are turning things to `None` above in some cases